File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    int shipWithinDays(vector<int>& weights, int days) {
        //Binary search
        
        //search space
        int low = 0, high = 0;
        for(int weight :weights) {
            low = max(low, weight);
            high += weight;
        }
        
        int ans = high;
        while(low <= high) {
            int maxWeightCanCarry = low +(high -low) /2;
            
            if(isPossibleToCarryWeight(weights, maxWeightCanCarry, days)) {
                ans = maxWeightCanCarry;
                high = maxWeightCanCarry -1;
            } else {
                low = maxWeightCanCarry +1;
            }
        }
        
        return ans;
    }
    
private:
    bool isPossibleToCarryWeight(vector<int>& weights, int maxWeightCanCarry, int days) {
        int day = 1;
        int curWeight = 0;
        
        for(int weight :weights) {
            if(weight +curWeight > maxWeightCanCarry) {
                day++;
                curWeight = 0;
            }
            curWeight += weight;
        }
        
        return (day <= days);
    }
};

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        
        int low = 0, high = 0;
        for(int weight :weights) {
            low = Math.max(low, weight);
            high += weight;
        }
        
        int minWeightReq = high;
        while(low <= high) {
            int mid = low +((high -low) /2);
            
            if(isEnoughWeight(weights, mid, days)) {
                minWeightReq = mid;
                high = mid -1;
            } else {
                low = mid +1;
            }
            
        }
        
        return minWeightReq;
    }
    
    private boolean isEnoughWeight(int[] weights, int maxCapacity, int days) {
        int currWeight = 0;
        
        for(int weight :weights) {
            if(currWeight +weight > maxCapacity) {
                days--;
                currWeight = 0;
            }
            currWeight += weight;
        }
        
        return (days > 0 && currWeight <= maxCapacity);
    }
    
}